presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,101 @@
+import java.util.Optional;
+
+import static com.facebook.presto.operator.scalar.GroupingOperationFunction.GROUPING;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public final class GroupingOperationRewriter
+{
+    private GroupingOperationRewriter() {}
+
+    public static Expression rewriteGroupingOperation(GroupingOperation expression, QuerySpecification queryNode, Analysis analysis, Metadata metadata, Optional<Symbol> groupIdSymbol)
+    {
+        requireNonNull(queryNode, "node is null");
+        requireNonNull(analysis, "analysis is null");
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(groupIdSymbol, "groupIdSymbol is null");
+
+        checkState(queryNode.getGroupBy().isPresent(), "GroupBy node must be present");
+
+        // No GroupIdNode and a GROUPING() operation imply a single grouping, which
+        // means that any columns specified as arguments to GROUPING() will be included
+        // in the group and none of them will be aggregated over. Hence, re-write the
+        // GroupingOperation to a constant literal of 0.
+        // See SQL:2011:4.16.2 and SQL:2011:6.9.10.
+        boolean isSimpleGroupByPresent = queryNode.getGroupBy().get().getGroupingElements().stream().anyMatch(SimpleGroupBy.class::isInstance);
+        if (isSimpleGroupByPresent) {
+            return new LongLiteral("0");
+        }
+        else {
+            checkState(groupIdSymbol.isPresent(), "groupId symbol is missing");
+
+            List<Expression> columnReferences = ImmutableList.copyOf(analysis.getColumnReferences());
+            List<Expression> groupingOrdinals = expression.getGroupingColumns().stream()
+                    .map(columnReferences::indexOf)
+                    .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                    .collect(toImmutableList());
+
+            List<List<Expression>> groupingSetOrdinals;
+            ImmutableList.Builder<List<Expression>> groupingSetOrdinalsBuilder = ImmutableList.builder();
+            for (List<Expression> groupingSet : analysis.getGroupingSets(queryNode)) {
+                groupingSetOrdinalsBuilder.add(groupingSet.stream()
+                        .map(columnReferences::indexOf)
+                        .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                        .collect(toImmutableList()));
+            }
+            groupingSetOrdinals = groupingSetOrdinalsBuilder.build();
+
+            Expression firstGroupingArgument = groupIdSymbol.get().toSymbolReference();
+            List<Expression> newGroupingArguments = ImmutableList.of(
+                    firstGroupingArgument,
```

Comment:
I think the style is to either have all arguments on newlines or none of them. So I'm going to keep this line as is. However, there are other violations of this style in this file which I did fix.